File: kyrie/src/main/java/com/github/alexjlockwood/kyrie/CircleNode.java
package com.github.alexjlockwood.kyrie;

import android.graphics.Path;
import android.graphics.RectF;
import android.support.annotation.FloatRange;
import android.support.annotation.NonNull;

import java.util.List;

/** A {@link Node} that paints a circle. */
public final class CircleNode extends RenderNode {
  @NonNull private final List<Animation<?, Float>> centerX;
  @NonNull private final List<Animation<?, Float>> centerY;
  @NonNull private final List<Animation<?, Float>> radius;

  private CircleNode(
      List<Animation<?, Float>> rotation,
      List<Animation<?, Float>> pivotX,
      List<Animation<?, Float>> pivotY,
      List<Animation<?, Float>> scaleX,
      List<Animation<?, Float>> scaleY,
      List<Animation<?, Float>> translateX,
      List<Animation<?, Float>> translateY,
      List<Animation<?, Integer>> fillColor,
      List<Animation<?, Float>> fillAlpha,
      List<Animation<?, Integer>> strokeColor,
      List<Animation<?, Float>> strokeAlpha,
      List<Animation<?, Float>> strokeWidth,
      List<Animation<?, Float>> trimPathStart,
      List<Animation<?, Float>> trimPathEnd,
      List<Animation<?, Float>> trimPathOffset,
      @StrokeLineCap int strokeLineCap,
      @StrokeLineJoin int strokeLineJoin,
      List<Animation<?, Float>> strokeMiterLimit,
      List<Animation<?, float[]>> strokeDashArray,
      List<Animation<?, Float>> strokeDashOffset,
      @FillType int fillType,
      boolean isStrokeScaling,
      List<Animation<?, Float>> centerX,
      List<Animation<?, Float>> centerY,
      List<Animation<?, Float>> radius) {
    super(
        rotation,
        pivotX,
        pivotY,
        scaleX,
        scaleY,
        translateX,
        translateY,
        fillColor,
        fillAlpha,
        strokeColor,
        strokeAlpha,
        strokeWidth,
        trimPathStart,
        trimPathEnd,
        trimPathOffset,
        strokeLineCap,
        strokeLineJoin,
        strokeMiterLimit,
        strokeDashArray,
        strokeDashOffset,
        fillType,
        isStrokeScaling);
    this.centerX = centerX;
    this.centerY = centerY;
    this.radius = radius;
  }

  @NonNull
  List<Animation<?, Float>> getCenterX() {
    return centerX;
  }

  @NonNull
  List<Animation<?, Float>> getCenterY() {
    return centerY;
  }

  @NonNull
  List<Animation<?, Float>> getRadius() {
    return radius;
  }

  // <editor-fold desc="Layer">

  @NonNull
  @Override
  CircleLayer toLayer(PropertyTimeline timeline) {
    return new CircleLayer(timeline, this);
  }

  private static class CircleLayer extends RenderLayer {
    @NonNull private final Property<Float> centerX;
    @NonNull private final Property<Float> centerY;
    @NonNull private final Property<Float> radius;

    private final RectF tempRect = new RectF();

    public CircleLayer(PropertyTimeline timeline, CircleNode node) {
      super(timeline, node);
      centerX = registerAnimatableProperty(node.getCenterX());
      centerY = registerAnimatableProperty(node.getCenterY());
      radius = registerAnimatableProperty(node.getRadius());
    }

    @Override
    public void onInitPath(Path outPath) {
      final float cx = centerX.getAnimatedValue();
      final float cy = centerY.getAnimatedValue();
      final float r = radius.getAnimatedValue();
      tempRect.set(cx - r, cy - r, cx + r, cy + r);
      outPath.addOval(tempRect, Path.Direction.CW);
    }
  }

  // </editor-fold>

  // <editor-fold desc="Builder">

  public static Builder builder() {
    return new Builder();
  }

  /** Builder class used to create {@link CircleNode}s. */
  public static final class Builder extends RenderNode.Builder<Builder> {
    @NonNull private final List<Animation<?, Float>> centerX = Node.asAnimations(0f);
    @NonNull private final List<Animation<?, Float>> centerY = Node.asAnimations(0f);
    @NonNull private final List<Animation<?, Float>> radius = Node.asAnimations(0f);

    private Builder() {}

    // Center X.

    public Builder centerX(float initialCenterX) {
      return replaceFirstAnimation(centerX, Node.asAnimation(initialCenterX));
    }

    @SafeVarargs
    public final Builder centerX(Animation<?, Float>... animations) {
      return replaceAnimations(centerX, animations);
    }

    public Builder centerX(List<Animation<?, Float>> animations) {
      return replaceAnimations(centerX, animations);
    }

    // Center Y.

    public Builder centerY(float initialCenterY) {
      return replaceFirstAnimation(centerY, Node.asAnimation(initialCenterY));
    }

    @SafeVarargs
    public final Builder centerY(Animation<?, Float>... animations) {
      return replaceAnimations(centerY, animations);
    }

    public Builder centerY(List<Animation<?, Float>> animations) {
      return replaceAnimations(centerY, animations);
    }

    // Radius.

    public Builder radius(@FloatRange(from = 0f) float initialRadius) {
      return replaceFirstAnimation(radius, Node.asAnimation(initialRadius));
    }

    @SafeVarargs
    public final Builder radius(Animation<?, Float>... animations) {
      return replaceAnimations(radius, animations);
    }

    public Builder radius(List<Animation<?, Float>> animations) {
      return replaceAnimations(radius, animations);
    }

    @NonNull
    @Override
    Builder self() {
      return this;
    }

    @NonNull
    public CircleNode build() {
      return new CircleNode(
          rotation,
          pivotX,
          pivotY,
          scaleX,
          scaleY,
          translateX,
          translateY,
          fillColor,
          fillAlpha,
          strokeColor,
          strokeAlpha,
          strokeWidth,
          trimPathStart,
          trimPathEnd,
          trimPathOffset,
          strokeLineCap,
          strokeLineJoin,
          strokeMiterLimit,
          strokeDashArray,
          strokeDashOffset,
          fillType,
          isScalingStroke,
          centerX,
          centerY,
          radius);
    }
  }

  // </editor-fold>
}

File: kyrie/src/main/java/com/github/alexjlockwood/kyrie/CircleNode.kt
package com.github.alexjlockwood.kyrie

import android.graphics.Path
import android.graphics.RectF
import androidx.annotation.FloatRange

/** A [Node] that paints a circle. */
class CircleNode private constructor(
        rotation: List<Animation<*, Float>>,
        pivotX: List<Animation<*, Float>>,
        pivotY: List<Animation<*, Float>>,
        scaleX: List<Animation<*, Float>>,
        scaleY: List<Animation<*, Float>>,
        translateX: List<Animation<*, Float>>,
        translateY: List<Animation<*, Float>>,
        fillColor: List<Animation<*, Int>>,
        fillColorComplex: ComplexColor?,
        fillAlpha: List<Animation<*, Float>>,
        strokeColor: List<Animation<*, Int>>,
        strokeColorComplex: ComplexColor?,
        strokeAlpha: List<Animation<*, Float>>,
        strokeWidth: List<Animation<*, Float>>,
        trimPathStart: List<Animation<*, Float>>,
        trimPathEnd: List<Animation<*, Float>>,
        trimPathOffset: List<Animation<*, Float>>,
        strokeLineCap: StrokeLineCap,
        strokeLineJoin: StrokeLineJoin,
        strokeMiterLimit: List<Animation<*, Float>>,
        strokeDashArray: List<Animation<*, FloatArray>>,
        strokeDashOffset: List<Animation<*, Float>>,
        fillType: FillType,
        isStrokeScaling: Boolean,
        private val centerX: List<Animation<*, Float>>,
        private val centerY: List<Animation<*, Float>>,
        private val radius: List<Animation<*, Float>>
) : RenderNode(
        rotation,
        pivotX,
        pivotY,
        scaleX,
        scaleY,
        translateX,
        translateY,
        fillColor,
        fillColorComplex,
        fillAlpha,
        strokeColor,
        strokeColorComplex,
        strokeAlpha,
        strokeWidth,
        trimPathStart,
        trimPathEnd,
        trimPathOffset,
        strokeLineCap,
        strokeLineJoin,
        strokeMiterLimit,
        strokeDashArray,
        strokeDashOffset,
        fillType,
        isStrokeScaling
) {

    // <editor-fold desc="Layer">

    override fun toLayer(timeline: PropertyTimeline): CircleLayer {
        return CircleLayer(timeline, this)
    }

    internal class CircleLayer(timeline: PropertyTimeline, node: CircleNode) : RenderNode.RenderLayer(timeline, node) {
        private val centerX = registerAnimatableProperty(node.centerX)
        private val centerY = registerAnimatableProperty(node.centerY)
        private val radius = registerAnimatableProperty(node.radius)

        private val tempRect = RectF()

        override fun onInitPath(outPath: Path) {
            val cx = centerX.animatedValue
            val cy = centerY.animatedValue
            val r = radius.animatedValue
            tempRect.set(cx - r, cy - r, cx + r, cy + r)
            outPath.addOval(tempRect, Path.Direction.CW)
        }
    }

    // </editor-fold>

    // <editor-fold desc="Builder">

    @DslMarker
    private annotation class CircleNodeMarker

    /** Builder class used to create [CircleNode]s. */
    @CircleNodeMarker
    class Builder internal constructor() : RenderNode.Builder<Builder>() {
        private val centerX = asAnimations(0f)
        private val centerY = asAnimations(0f)
        private val radius = asAnimations(0f)

        // Center X.

        fun centerX(initialCenterX: Float): Builder {
            return replaceFirstAnimation(centerX, asAnimation(initialCenterX))
        }

        @SafeVarargs
        fun centerX(vararg animations: Animation<*, Float>): Builder {
            return replaceAnimations(centerX, *animations)
        }

        fun centerX(animations: List<Animation<*, Float>>): Builder {
            return replaceAnimations(centerX, animations)
        }

        // Center Y.

        fun centerY(initialCenterY: Float): Builder {
            return replaceFirstAnimation(centerY, asAnimation(initialCenterY))
        }

        @SafeVarargs
        fun centerY(vararg animations: Animation<*, Float>): Builder {
            return replaceAnimations(centerY, *animations)
        }

        fun centerY(animations: List<Animation<*, Float>>): Builder {
            return replaceAnimations(centerY, animations)
        }

        // Radius.

        fun radius(@FloatRange(from = 0.0) initialRadius: Float): Builder {
            return replaceFirstAnimation(radius, asAnimation(initialRadius))
        }

        @SafeVarargs
        fun radius(vararg animations: Animation<*, Float>): Builder {
            return replaceAnimations(radius, *animations)
        }

        fun radius(animations: List<Animation<*, Float>>): Builder {
            return replaceAnimations(radius, animations)
        }

        override val self = this

        override fun build(): CircleNode {
            return CircleNode(
                    rotation,
                    pivotX,
                    pivotY,
                    scaleX,
                    scaleY,
                    translateX,
                    translateY,
                    fillColor,
                    fillColorComplex,
                    fillAlpha,
                    strokeColor,
                    strokeColorComplex,
                    strokeAlpha,
                    strokeWidth,
                    trimPathStart,
                    trimPathEnd,
                    trimPathOffset,
                    strokeLineCap,
                    strokeLineJoin,
                    strokeMiterLimit,
                    strokeDashArray,
                    strokeDashOffset,
                    fillType,
                    isScalingStroke,
                    centerX,
                    centerY,
                    radius
            )
        }
    }

    // </editor-fold>

    companion object {

        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }
    }
}
